src/librustc/infer/error_reporting/anon_anon_conflict.rs

```diff
@@ -39,54 +47,84 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+                    }
                 }
+
             } else {
                 return false;
             }
         } else {
-            return false; // inapplicable
+            return false; //inapplicable
         };
 
-        if let (Some(sup_arg), Some(sub_arg)) =
+        let (label1, label2) = if let (Some(sup_arg), Some(sub_arg)) =
             (self.find_arg_with_anonymous_region(sup, sup),
              self.find_arg_with_anonymous_region(sub, sub)) {
             let ((anon_arg1, _, _, _), (anon_arg2, _, _, _)) = (sup_arg, sub_arg);
-
             let span_label_var1 = if let Some(simple_name) = anon_arg1.pat.simple_name() {
-                format!(" from `{}` ", simple_name)
+                format!(" from `{}`", simple_name)
             } else {
-                format!(" ")
+                format!("")
             };
 
             let span_label_var2 = if let Some(simple_name) = anon_arg2.pat.simple_name() {
-                format!(" into `{}` ", simple_name)
+                format!(" into `{}`", simple_name)
             } else {
-                format!(" ")
+                format!("")
             };
 
+            (span_label_var1, span_label_var2)
+        } else {
+            return false;
+        };
+
+        if let Some(error_label) =
+            self.process_anon_anon_conflict(first_is_struct, second_is_struct) {
             struct_span_err!(self.tcx.sess, span, E0623, "lifetime mismatch")
                 .span_label(ty1.span,
-                            format!("these references are not declared with the same lifetime..."))
+                            format!("{} are declared with different lifetimes...", error_label))
                 .span_label(ty2.span, format!(""))
-                .span_label(span,
-                            format!("...but data{}flows{}here", span_label_var1, span_label_var2))
+                .span_label(span, format!("...but data{} flows{} here", label1, label2))
                 .emit();
+            return true;
         } else {
             return false;
         }
+    }
 
-        return true;
+    // try to pre-process the errors, for the purpose of generating different
+    // span labels for the different combinations of the two parameters which can be
+    // either references or structs.
+    fn process_anon_anon_conflict(&self,
+                                  is_arg1_struct: bool,
+                                  is_arg2_struct: bool)
+                                  -> Option<(String)> {
+        let arg1_label = {
+            if is_arg1_struct && is_arg2_struct {
+                format!("these structs")
+            } else if is_arg1_struct && !is_arg2_struct {
+                format!("the struct and reference")
+            } else if !is_arg1_struct && is_arg2_struct {
+                format!("the reference and the struct")
+            } else {
+                format!("these references")
+            }
+        };
+        Some(arg1_label)
```

Comment:
There's no case where this method doesn't return `Some`, so you can change the method's signature to return `String`, instead of `Option<String>`.